[sandbox] Add initial support for windows sandbox

rongjiecomputer opened the pull request.

rongjiecomputer commented on src/main/java/com/google/devtools/build/lib/exec/local/WindowsLocalEnvProvider.java:
```diff
@@ -30,7 +30,7 @@
    * @param clientEnv a map of the current Bazel command's environment
    */
-  WindowsLocalEnvProvider(Map<String, String> clientEnv) {
+  public WindowsLocalEnvProvider(Map<String, String> clientEnv) {
```
Needed so that `WindowsSandboxedSpawnRunner` can use it.

rongjiecomputer commented on src/main/java/com/google/devtools/build/lib/sandbox/WindowsSandboxedSpawn.java:
```diff
@@ -0,0 +1,64 @@
+/**
+ * Creates an execRoot for a Spawn by copying input files.
+ */
+public class WindowsSandboxedSpawn extends AbstractContainerizingSandboxedSpawn {
```
This does the same thing as `CopyingSandboxedSpawn` for now as I just want to get things to work first.

rongjiecomputer replied:
> Do you mean it still copies instead of something else?

Yes

> What will you replace copying with?

I was thinking to collect input file paths from `copyFile` method below so that I can mark them as readable in windows-sandbox. To do that, this class will hold a reference to the commandlinebuilder, which now I think is a horrible way to do stuff.

> Thinking more, what is the purpose of a unique execroot? I suppose it (a) allows staging the declared inputs only, and (b) it isolates actions so they don't overwrite each other's intermediate outputs.
I understand (b) is useful, but it seems (a) isn't required for the detours-based sandbox.

You are right. (a) isn't required for the detours-based sandbox. In the final stage, the sandbox will just run in the common execroot, files will be marked readable or read-write explicitly. (b) is nice to have, but not in the scope for now.

I just found out about `SandboxHelpers.processInputFiles` which seems to be what I need, so I think this is what I am going to do:

- Delete `WindowsSandboxedSpawn.java` this class and use `CopyingSandboxedSpawn` for this PR.
- Figure out how to use ``SandboxHelpers.processInputFiles` and use that when building windows-sandbox command line.
- Create a `DummySandboxedSpawn` for windows-sandbox which does nothing.

laszlocsomor replied:
SGTM, thank you!

rongjiecomputer commented:
/cc @laszlocsomor

laszlocsomor commented on src/main/java/com/google/devtools/build/lib/sandbox/WindowsSandboxedSpawnRunner.java:
```diff
@@ -0,0 +1,145 @@
+    // the normal execroot does.
+    Path sandboxExecRoot = sandboxPath.getRelative("execroot").getRelative(execRoot.getBaseName());
+    sandboxExecRoot.getParentDirectory().createDirectory();
+    sandboxExecRoot.createDirectory();
```
same here

laszlocsomor commented on src/main/java/com/google/devtools/build/lib/sandbox/WindowsSandboxedSpawnRunner.java:
```diff
@@ -0,0 +1,145 @@
+    Path sandboxPath =
+        sandboxBase.getRelative(getName()).getRelative(Integer.toString(context.getId()));
+    sandboxPath.getParentDirectory().createDirectory();
+    sandboxPath.createDirectory();
```
You can combine these 2 lines as `sandboxPath.createDirectoryAndParents()`.

laszlocsomor commented on src/main/java/com/google/devtools/build/lib/sandbox/WindowsSandboxedSpawnRunner.java:
```diff
@@ -0,0 +1,145 @@
+    sandboxExecRoot.createDirectory();
+
+    Map<String, String> environment =
+        localEnvProvider.rewriteLocalEnv(spawn.getEnvironment(), binTools, "/tmp");
```
`/tmp` should be a Windows path I think

rongjiecomputer replied:
About this, I don't really know what is the good path I should set. `C:\Windows\Temp` is writable without admin permission, but not readable without admin permission. `%USERPROFILE%\AppData\Temp` probably exists, but `USERPROFILE` env var might not be present to construct the path.

In `LocalSpawnRunner`, it creates a unique tmp dir for this: https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/exec/local/LocalSpawnRunner.java;drc=c26e3396cb341e173b4be5c780cafac3ec386df5;l=310 . Should I do the same here?

laszlocsomor replied:
Don't use any of those. If neither TEMP nor TMP are set, things like `java.nio.file.Files.createTempDirectory` break anyway.

> Should I do the same here?

Yes please.

laszlocsomor commented on src/main/java/com/google/devtools/build/lib/sandbox/AbstractSandboxSpawnRunner.java:
```diff
@@ -276,7 +289,11 @@ private void addWritablePath(
       // the symlink itself writable, not what it points to. Therefore we need to resolve symlinks
       // in `path`, however for that we need `path` to exist.
-      writablePaths.add(path.resolveSymbolicLinks());
+      if (path.isSymbolicLink()) {
```
Thanks!

laszlocsomor commented on src/main/java/com/google/devtools/build/lib/sandbox/AbstractSandboxSpawnRunner.java:
```diff
@@ -238,23 +239,35 @@ private boolean wasTimeout(Duration timeout, Duration wallTime) {
       Path path = fileSystem.getPath(writablePath);
       writablePaths.add(path);
-      writablePaths.add(path.resolveSymbolicLinks());
+      if (path.isSymbolicLink()) {
```
thanks

rongjiecomputer commented:
> Is it possible to integration-test this already? If so, please add test(s).

After spending porting `//src/test/shell/bazel:bazel_sandboxing_cpp_test` for Windows, I realized that there are still quite a few bugs I need to fix in windows-sandbox side. Can we land this PR first?

Also, it seems like the tests under `//src/test/shell/` does not use "Bazel that is built in `bazel-bin`" but instead uses the "main Bazel that is building everything". I don't think this is intended.  I noticed this because I was using an older Bazel that does not have the new sandbox flags I introduced and the new tests I added complain it cannot recognized the flags.

laszlocsomor commented:
> > Is it possible to integration-test this already? If so, please add test(s).
> 
> After spending porting `//src/test/shell/bazel:bazel_sandboxing_cpp_test` for Windows, I realized that there are still quite a few bugs I need to fix in windows-sandbox side. Can we land this PR first?

Ok, let's land it first.

> Also, it seems like the tests under `//src/test/shell/` does not use "Bazel that is built in `bazel-bin`" but instead uses the "main Bazel that is building everything". I don't think this is intended. I noticed this because I was using an older Bazel that does not have the new sandbox flags I introduced and the new tests I added complain it cannot recognized the flags.

That sounds very unlikely to me, it'd beat the purpose of those tests. Those tests should be using bazel built from HEAD. Are you certain they use the installed bazel version?

rongjiecomputer commented:
> That sounds very unlikely to me, it'd beat the purpose of those tests. Those tests should be using bazel built from HEAD. Are you certain they use the installed bazel version?

This was what happened:

- Use old Bazel to build Bazel HEAD and run the test: test complains it cannot recognize `--experimental_use_windows_sandbox=yes`
- Use Bazel HEAD to build Bazel HEAD (for the first run, it almost completely rebuilds everything), test finally recognize the flag and does stuffs that I expect it to do.

laszlocsomor commented:
> > That sounds very unlikely to me, it'd beat the purpose of those tests. Those tests should be using bazel built from HEAD. Are you certain they use the installed bazel version?
> 
> This was what happened:
> 
> * Use old Bazel to build Bazel HEAD and run the test: test complains it cannot recognize `--experimental_use_windows_sandbox=yes`
> * Use Bazel HEAD to build Bazel HEAD (for the first run, it almost completely rebuilds everything), test finally recognize the flag and does stuffs that I expect it to do.

That sounds like a bug. Could you file a bug please and /cc me?

laszlocsomor commented on src/main/java/com/google/devtools/build/lib/sandbox/WindowsSandboxedSpawnRunner.java:
```diff
@@ -0,0 +1,154 @@
+/** Spawn runner that uses BuildXL Sandbox APIs to execute a local subprocess. */
+final class WindowsSandboxedSpawnRunner extends AbstractSandboxSpawnRunner {
+
+  private final FileSystem fileSystem;
```
static analyzer says, this field is never used

laszlocsomor commented on src/main/java/com/google/devtools/build/lib/sandbox/WindowsSandboxedSpawnRunner.java:
```diff
@@ -0,0 +1,154 @@
+final class WindowsSandboxedSpawnRunner extends AbstractSandboxSpawnRunner {
+
+  private final FileSystem fileSystem;
+  private final BlazeDirectories blazeDirs;
```
static analyzer says, this field is never used

laszlocsomor commented:
Ah, apparently the analyzer said they are never **read**, they were assigned. Sorry, my mistake.

Removing the assignments fixes that.

The PR seems to break an internal test but I'm not sure why. Investigating.

laszlocsomor commented on src/main/java/com/google/devtools/build/lib/sandbox/WindowsSandboxedSpawnRunner.java:
```diff
@@ -0,0 +1,152 @@
+    return runSpawn(spawn, sandbox, context, execRoot, timeout, statisticsPath);
+  }
+
+  private Path createActionTemp(Path execRoot) throws IOException {
```
can be `static`

laszlocsomor commented on src/main/java/com/google/devtools/build/lib/sandbox/AbstractSandboxSpawnRunner.java:
```diff
@@ -238,23 +239,36 @@ private boolean wasTimeout(Duration timeout, Duration wallTime) {
       writablePaths.add(path);
-      writablePaths.add(path.resolveSymbolicLinks());
+      // TODO(laszlocsomor) Remove if guard when path.resolveSymbolicLinks supports non-symlink
+      if (path.isSymbolicLink()) {
```
> The PR seems to break an internal test but I'm not sure why. Investigating.

This new if-condition, and the one I commented on below, are responsible.

This workaround will do for now:

```diff
-	      if (path.isSymbolicLink()) {
+	      // TODO(laszlocsomor): figure out why OS.getCurrent() != OS.WINDOWS is required, and remove it
+	      if (OS.getCurrent() != OS.WINDOWS || path.isSymbolicLink()) {
```

laszlocsomor commented on src/main/java/com/google/devtools/build/lib/sandbox/AbstractSandboxSpawnRunner.java:
```diff
@@ -276,7 +290,13 @@ private void addWritablePath(
-      writablePaths.add(path.resolveSymbolicLinks());
+      //
+      // TODO(laszlocsomor) Remove if guard when path.resolveSymbolicLinks supports non-symlink
+      if (path.isSymbolicLink()) {
```
And here.

laszlocsomor commented on src/main/java/com/google/devtools/build/lib/sandbox/SandboxModule.java:
```diff
@@ -272,9 +272,9 @@ private void setup(CommandEnvironment cmdEnv, ExecutorBuilder builder)
+    boolean windowsSandboxSupported;
     try (SilentCloseable c = Profiler.instance().profile("shouldUseWindowsSandbox")) {
-      useWindowsSandbox = shouldUseWindowsSandbox(options.useWindowsSandbox, windowsSandboxPath);
+      windowsSandboxSupported = shouldUseWindowsSandbox(options.useWindowsSandbox, windowsSandboxPath);
```
To my surprise, checking `OS.getCurrent()` doesn't seem to be needed in `shouldUseWindowsSandbox()`. Do you know why?

rongjiecomputer replied:
`options.useWindowsSandbox` is by default `no`, so this function will just return `false` without doing anything.
